File: server/src/plugins/runtime/node/proxy/deviceManager.ts
import { NamespaceManager } from '../../../../rpc/namespaces.js';
import { CameraDeviceProxy } from './cameraDevice.js';

import type { Logger } from '@camera.ui/common';
import type { Promisify, RPCClient } from '@camera.ui/rpc';
import type { BasePlugin, Camera, DeviceManager, DiscoveredCamera, PluginInfo } from '@camera.ui/sdk';
import type { DeviceManagerInterface, DeviceManagerListenerMessagePayload } from '../../../../rpc/interfaces/device.js';
import type { DiscoveryManagerInterface } from '../../../../rpc/interfaces/discovery.js';
import type { DeviceManagerNamespaces, DiscoveryManagerNamespaces, PluginNamespaces } from '../../../../rpc/namespaces.js';
import type { StorageController } from '../storageController.js';

export class DeviceManagerProxy implements DeviceManager {
  #proxy: RPCClient;
  #storageController: StorageController;
  #logger: Logger;
  #pluginInfo: PluginInfo;
  #pluginInstance?: BasePlugin;

  #initialized = false;

  #closeRequests?: () => void;
  #namespaces: DeviceManagerNamespaces & PluginNamespaces & DiscoveryManagerNamespaces;

  #devices = new Map<string, CameraDeviceProxy>();

  constructor(proxy: RPCClient, storageController: StorageController, pluginInfo: PluginInfo, logger: Logger) {
    this.#proxy = proxy;
    this.#storageController = storageController;
    this.#logger = logger;
    this.#pluginInfo = pluginInfo;
    this.#namespaces = {
      ...NamespaceManager.deviceManagerNamespaces(),
      ...NamespaceManager.pluginNamespaces(this.#pluginInfo.id),
      ...NamespaceManager.discoveryManagerNamespaces(),
    };
  }

  public setPlugin(plugin: BasePlugin): void {
    this.#pluginInstance = plugin;
  }

  public async init(): Promise<void> {
    if (this.#initialized) {
      return;
    }

    this.#initialized = true;
    this.#closeRequests = await this.#proxy.onRequest<DeviceManagerListenerMessagePayload>(this.#namespaces.pluginDeviceManagerSubject, this.#onEventMessage.bind(this));
  }

  public async getCamera(cameraIdOrName: string): Promise<CameraDeviceProxy | undefined> {
    let cameraDevice = await this.#getCameraDevice(cameraIdOrName);

    if (!cameraDevice) {
      const camera = await this.#deviceManagerProxy.getCamera(cameraIdOrName, this.#pluginInfo.id);

      if (camera) {
        cameraDevice = await this.#getCameraDevice(camera);
      }
    }

    return cameraDevice;
  }

  public async pushDiscoveredCameras(cameras: DiscoveredCamera[]): Promise<void> {
    await this.#discoveryManagerProxy.pushDiscoveredCameras(this.#pluginInfo.id, cameras);
  }

  public async configureCameras(cameraDevices: CameraDeviceProxy[]): Promise<void> {
    await Promise.all(cameraDevices.map((cameraDevice) => this.#getCameraDevice(cameraDevice)));
  }

  /** Internal method to close the device manager and clean up resources */
  public async close(): Promise<void> {
    this.#initialized = false;
    this.#closeRequests?.();

    for (const device of this.#devices.values()) {
      await device.cleanup();
    }
    this.#devices.clear();
  }

  get #deviceManagerProxy(): Promisify<DeviceManagerInterface> {
    return this.#proxy.createProxy<DeviceManagerInterface>(this.#namespaces.deviceManagerRpc);
  }

  get #discoveryManagerProxy(): Promisify<DiscoveryManagerInterface> {
    return this.#proxy.createProxy<DiscoveryManagerInterface>(this.#namespaces.discoveryManagerRpc);
  }

  async #onEventMessage(event: DeviceManagerListenerMessagePayload): Promise<void> {
    if (!this.#pluginInstance) {
      this.#logger.warn('Plugin instance not set, cannot handle lifecycle event');
      return;
    }

    switch (event.type) {
      case 'cameraAdded': {
        const { camera } = event.data as { camera: Camera };
        if (this.#devices.has(camera._id)) {
          break;
        }
        const cameraDevice = await this.#getCameraDevice(camera);
        await this.#pluginInstance.onCameraAdded(cameraDevice);
        break;
      }
      case 'cameraReleased': {
        const { cameraId } = event.data as { cameraId: string };
        await this.#pluginInstance.onCameraReleased(cameraId);
        const cameraDevice = this.#devices.get(cameraId);
        await cameraDevice?.cleanup();
        await this.#removeCameraStorage(cameraId);
        this.#devices.delete(cameraId);
        break;
      }
    }
  }

  #getCameraDevice(camera: Camera): Promise<CameraDeviceProxy>;
  #getCameraDevice(cameraDevice: CameraDeviceProxy): Promise<CameraDeviceProxy>;
  #getCameraDevice(identifier: string): Promise<CameraDeviceProxy | undefined>;
  async #getCameraDevice(cameraOrIdentifier: Camera | CameraDeviceProxy | string): Promise<CameraDeviceProxy | undefined> {
    let cameraDevice: CameraDeviceProxy | undefined;

    if (typeof cameraOrIdentifier === 'string') {
      cameraDevice = Array.from(this.#devices.values()).find((cameraDevice) => cameraDevice.id === cameraOrIdentifier || cameraDevice.name === cameraOrIdentifier);
    } else if (cameraOrIdentifier instanceof CameraDeviceProxy) {
      cameraDevice = cameraOrIdentifier;

      if (this.#devices.has(cameraDevice.id)) {
        cameraDevice = this.#devices.get(cameraDevice.id)!;
      } else {
        this.#devices.set(cameraDevice.id, cameraDevice);
      }
    } else {
      const camera = cameraOrIdentifier;

      if (this.#devices.has(camera._id)) {
        cameraDevice = this.#devices.get(camera._id)!;
      } else {
        const cameraLogger = this.#logger.createLogger({ suffix: camera.name, targetId: camera._id, targetType: 'camera' });
        cameraDevice = new CameraDeviceProxy(this.#proxy, this.#storageController, camera, this.#pluginInfo, cameraLogger);
        this.#devices.set(camera._id, cameraDevice);
      }
    }

    if (cameraDevice) {
      await this.#createCameraStorage(cameraDevice.id);
      await cameraDevice?.init();
    }

    return cameraDevice;
  }

  async #createCameraStorage(cameraId: string): Promise<void> {
    await this.#storageController.createStorage('camera', cameraId);
  }

  async #removeCameraStorage(cameraId: string): Promise<void> {
    await this.#storageController.removeStorage('camera', cameraId);
  }
}

File: server/src/plugins/runtime/python/proxy/device_manager.py
from __future__ import annotations

import asyncio
from typing import Any

from _camera_ui_tools.camera_ui_common import (
    LoggerService,
)
from _camera_ui_tools.camera_ui_rpc import CloseHandler, RPCClient
from _camera_ui_tools.camera_ui_sdk import (
    BasePlugin,
    Camera,
    DeviceManager,
    DiscoveredCamera,
)
from plugins.runtime.python.namespaces import (
    DeviceManagerNamespaces,
    DiscoveryManagerNamespaces,
    NamespaceManager,
    PluginNamespaces,
)
from plugins.runtime.python.rpc.typings import (
    DeviceManagerInterface,
    DiscoveryManagerInterface,
)
from plugins.runtime.python.storage_controller import StorageController
from plugins.runtime.python.typings import PluginInfo

from .camera_device import CameraDeviceProxy


class DeviceManagerProxy(DeviceManager):
    def __init__(
        self,
        proxy: RPCClient,
        storage_controller: StorageController,
        logger: LoggerService,
        plugin: PluginInfo,
    ):
        self.__initialized = False
        self.__plugin_instance: BasePlugin | None = None

        self.__proxy = proxy
        self.__storage_controller = storage_controller
        self.__logger = logger
        self.__plugin = plugin
        self.__close_request: CloseHandler | None = None

        self.__devices: dict[str, CameraDeviceProxy] = {}
        self.__namespaces: tuple[DeviceManagerNamespaces, PluginNamespaces, DiscoveryManagerNamespaces] = (
            NamespaceManager.device_manager_namespaces(),
            NamespaceManager.plugin_namespaces(self.__plugin["id"]),
            NamespaceManager.discovery_manager_namespaces(),
        )

    @property
    def __device_manager_proxy(self) -> DeviceManagerInterface:
        return self.__proxy.create_proxy(self.__namespaces[0].device_manager_rpc, DeviceManagerInterface)

    @property
    def __discovery_manager_proxy(self) -> DiscoveryManagerInterface:
        return self.__proxy.create_proxy(
            self.__namespaces[2].discovery_manager_rpc, DiscoveryManagerInterface
        )

    def set_plugin(self, plugin: BasePlugin) -> None:
        self.__plugin_instance = plugin

    async def init(self) -> None:
        if self.__initialized:
            return

        self.__initialized = True
        self.__close_request = await self.__proxy.on_request(
            self.__namespaces[1].plugin_device_manager_subject, self.__on_event_message
        )

    async def getCamera(self, cameraIdOrName: str) -> CameraDeviceProxy | None:
        camera_device = await self.__get_camera_device(cameraIdOrName)

        if not camera_device:
            camera = await self.__device_manager_proxy.getCamera(cameraIdOrName, self.__plugin["id"])

            if camera:
                camera_device = await self.__get_camera_device(camera)

        return camera_device

    async def pushDiscoveredCameras(self, cameras: list[DiscoveredCamera]) -> None:
        await self.__discovery_manager_proxy.pushDiscoveredCameras(self.__plugin["id"], cameras)

    async def configureCameras(self, camera_devices: list[CameraDeviceProxy]) -> None:
        await asyncio.gather(*[self.__get_camera_device(camera_device) for camera_device in camera_devices])

    async def close(self) -> None:
        """Internal method to close the device manager proxy and cleanup resources."""
        self.__initialized = False
        if self.__close_request:
            await self.__close_request()

        for device in self.__devices.values():
            await device.cleanup()
        self.__devices.clear()

    async def __on_event_message(self, event: Any) -> None:
        if not self.__plugin_instance:
            self.__logger.warn("Plugin instance not set, cannot handle lifecycle event")
            return

        event_type = event.get("type")
        data = event.get("data", {})

        if event_type == "cameraAdded":
            camera: Camera = data.get("camera")
            camera_device = await self.__get_camera_device(camera)

            if camera_device:
                # Call plugin lifecycle callback
                await self.__plugin_instance.onCameraAdded(camera_device)

        elif event_type == "cameraReleased":
            camera_id: str = data.get("cameraId")

            # Call plugin lifecycle callback
            await self.__plugin_instance.onCameraReleased(camera_id)

            # Cleanup
            camera_device = self.__devices.get(camera_id)
            if camera_device:
                await camera_device.cleanup()

            await self.__remove_camera_storage(camera_id)

            if camera_id in self.__devices:
                del self.__devices[camera_id]

    async def __get_camera_device(
        self, camera_or_id: Camera | CameraDeviceProxy | str
    ) -> CameraDeviceProxy | None:
        camera_device: CameraDeviceProxy | None = None

        if isinstance(camera_or_id, str):
            id = camera_or_id

            camera_device = next(
                (device for device in self.__devices.values() if device.id == id or device.name == id),
                None,
            )
        elif isinstance(camera_or_id, CameraDeviceProxy):
            camera_device = camera_or_id
            if camera_device.id in self.__devices:
                camera_device = self.__devices[camera_device.id]
            else:
                self.__devices[camera_device.id] = camera_device
        else:
            camera = camera_or_id
            if camera["_id"] in self.__devices:
                camera_device = self.__devices[camera["_id"]]
            else:
                camera_logger = self.__logger.create_logger(
                    {
                        "suffix": camera["name"],
                        "target_id": camera["_id"],
                        "target_type": "camera",
                    }
                )
                camera_device = CameraDeviceProxy(
                    self.__proxy,
                    self.__storage_controller,
                    camera,
                    self.__plugin,
                    camera_logger,
                )

                self.__devices[camera["_id"]] = camera_device

        if camera_device:
            await self.__create_camera_storage(camera_device.id)
            await camera_device.init()

        return camera_device

    async def __create_camera_storage(self, camera_id: str) -> None:
        await self.__storage_controller.createStorage("camera", camera_id)

    async def __remove_camera_storage(self, camera_id: str) -> None:
        await self.__storage_controller.removeStorage("camera", camera_id)
